File: 单刷2023B/code/fig_geometry.m
% 2023B第一问 截面几何示意图 (2D) — 修订版
% 修正: 仅保留60°半开角/α弧在坡面与水平面交角处/垂距D标注/
%       几何元素对齐原点/左右深度标注/字体加大
% 输出: ../fig/fig_geometry.png
% 用法: matlab -batch "fig_geometry"
function fig_geometry
    out = fullfile(fileparts(mfilename('fullpath')), '..', 'fig', 'fig_geometry.png');

    D = 70.0; ALPHA = deg2rad(1.5); t30 = tan(pi/6);

    fig = figure('Position', [100 100 1050 520], 'Color', 'w');
    ax = axes('Parent', fig); hold(ax, 'on');
    axis(ax, 'equal');

    % ===== 海平面 =====
    plot(ax, [-165 165], [0 0], '-', 'Color', [0.12 0.31 0.47], 'LineWidth', 2.2);
    text(ax, 163, 8, '海平面', 'HorizontalAlignment', 'right', ...
        'Color', [0.12 0.31 0.47], 'FontSize', 14);

    % ===== 海底坡面 (y>0为坡下, 右深左浅) =====
    y = linspace(-160, 160, 200);
    yp = -(D + y * tan(ALPHA));
    plot(ax, y, yp, '-', 'Color', [0.55 0.27 0.07], 'LineWidth', 2.5);
    patch(ax, [y fliplr(y)], [yp fliplr(-130*ones(size(y)))], ...
        [0.82 0.71 0.55], 'FaceAlpha', 0.3, 'EdgeColor', 'none');

    % ===== 船 (原点) 与垂线 =====
    plot(ax, 0, 0, 'ks', 'MarkerSize', 12, 'MarkerFaceColor', 'k');
    text(ax, 12, 6, '船', 'FontSize', 14, 'FontWeight', 'bold');
    plot(ax, [0 0], [0 -D], 'k--', 'LineWidth', 1.2);

    % ===== 波束边缘 (与竖直法线成60°, 即半开角) =====
    for s = [-1 1]
        yy = linspace(0, s*130, 100);
        plot(ax, yy, -abs(yy)*t30, '-', 'Color', [0.18 0.46 0.71], ...
            'LineWidth', 1.8);
    end

    % ===== 交点 P_L / P_R =====
    yL = -D/(t30 + tan(ALPHA)); yR = D/(t30 - tan(ALPHA));
    zL = D + yL*tan(ALPHA); zR = D + yR*tan(ALPHA);
    plot(ax, [yL yR], [-zL -zR], 'ro', 'MarkerSize', 7, 'MarkerFaceColor', 'r');
    text(ax, yL-45, -zL-18, 'P_L', 'Color', 'r', 'FontWeight', 'bold', 'FontSize', 14);
    text(ax, yR+6, -zR-18, 'P_R', 'Color', 'r', 'FontWeight', 'bold', 'FontSize', 14);

    % ===== 覆盖宽度 W (海平面基准) =====
    plot(ax, [yL yR], [12 12], 'r-', 'LineWidth', 1.8);
    plot(ax, [yL yL], [8 16], 'r-', 'LineWidth', 1.8);
    plot(ax, [yR yR], [8 16], 'r-', 'LineWidth', 1.8);
    text(ax, (yL+yR)/2, 28, sprintf('W = %.1f m', yR-yL), ...
        'HorizontalAlignment', 'center', 'Color', 'r', ...
        'FontSize', 15, 'FontWeight', 'bold');

    % ===== 唯一角度: 60° (半开角, 波束边缘与竖直法线夹角, 以船为圆心) =====
    th = linspace(deg2rad(-90), deg2rad(-30), 40);
    r = 20;
    plot(ax, r*cos(th), r*sin(th), '-', 'Color', [0.18 0.46 0.71], ...
        'LineWidth', 1.5);
    text(ax, r*1.35*cos(deg2rad(-60)), r*1.35*sin(deg2rad(-60)), '60°', ...
        'HorizontalAlignment', 'center', 'Color', [0.18 0.46 0.71], ...
        'FontSize', 14);

    % ===== α角: 坡面与水平面交角 (圆心=船正下方海底点) =====
    th = linspace(pi, pi - deg2rad(1.5), 30);
    r = 14;
    plot(ax, r*cos(th), -D + r*sin(th), '-', 'Color', [0.44 0.19 0.63], ...
        'LineWidth', 1.5);
    text(ax, r*1.4*cos(pi - deg2rad(0.75)), -D + r*1.4*sin(pi - deg2rad(0.75)), ...
        '\alpha=1.5°', 'Color', [0.44 0.19 0.63], 'FontSize', 14);

    % ===== 垂距 D 标注 (船正下方, 虚线法线处) =====
    plot(ax, [10 10], [0 -D], 'k-', 'LineWidth', 1.2);
    plot(ax, [7 13], [0 0], 'k-', 'LineWidth', 1.2);
    plot(ax, [7 13], [-D -D], 'k-', 'LineWidth', 1.2);
    text(ax, 16, -D/2, '垂距 D=70m', 'FontSize', 13, ...
        'VerticalAlignment', 'middle');

    % ===== 左右端深度标注 (增强坡度视觉: 右深左浅) =====
    text(ax, -160, -(D-160*tan(ALPHA)) - 10, '浅 ~65.8m', ...
        'FontSize', 12, 'Color', [0.3 0.3 0.3], 'HorizontalAlignment', 'left');
    text(ax, 160, -(D+160*tan(ALPHA)) - 10, '深 ~74.2m', ...
        'FontSize', 12, 'Color', [0.3 0.3 0.3], 'HorizontalAlignment', 'right');

    xlim([-165 165]); ylim([-110 60]);
    xlabel('y (m)  向坡下为正', 'FontSize', 13);
    ylabel('z (m)  深度(向下)', 'FontSize', 13);
    grid(ax, 'on');
    hold(ax, 'off');

    exportgraphics(fig, out, 'Resolution', 300);
    fprintf('[OK] %s\n', out);
end

File: 单刷2023B/code/fig_geometry.py
"""
用途: 2023B多波束测线 截面几何示意图 (Matplotlib)
输出: output/fig_geometry.png
修正: z向下为正, 图上用 -z 绘制, 海平面在上, 海底在下
调用: python models/mbs/fig_geometry.py
"""
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.patches import Arc

plt.rcParams['font.sans-serif'] = ['SimHei', 'Microsoft YaHei']
plt.rcParams['axes.unicode_minus'] = False

alpha = np.deg2rad(1.5)   # 坡面倾角
D = 70.0                  # 中心水深
t30 = np.tan(np.deg2rad(30))   # 波束边缘与水平面夹角 30°

# 水平坐标 y (向坡下为正), 竖直绘制坐标 yp = -z (向下为深)
y = np.linspace(-160, 160, 200)
z_slope = D + y * np.tan(alpha)   # 真实深度 (向下)
yp_slope = -z_slope               # 绘图坐标 (海平面=0, 向下为负)

fig, ax = plt.subplots(figsize=(10.5, 5.2))
ax.set_aspect('equal')

# ===== 海平面 (yp=0) =====
ax.axhline(0, color='#1f4e79', lw=2.2)
ax.text(163, 8, '海平面', ha='right', color='#1f4e79', fontsize=11)

# ===== 海底坡面 (下方) =====
ax.plot(y, yp_slope, color='#8b4513', lw=2.5, label='海底坡面 (α=1.5°)')
ax.fill_between(y, yp_slope, -130, color='#d2b48c', alpha=0.3)

# ===== 船 (海平面上 yp=0) =====
ax.plot(0, 0, 'ks', ms=11, zorder=5)
ax.annotate('船', (0, 0), xytext=(10, 4), fontsize=11, fontweight='bold')

# ===== 波束边缘射线 (向下, 与竖直成60°) =====
for sign in (-1, 1):
    yy = np.linspace(0, sign * 130, 100)
    zz = -np.abs(yy) * t30        # 向下: yp = -|y|·tan30°
    ax.plot(yy, zz, color='#2e75b6', lw=1.8, alpha=0.85)
# 船到海底垂线 (水深)
ax.plot([0, 0], [0, -D], 'k--', lw=1.2, alpha=0.6)

# ===== 交点 =====
yL = -D / (t30 + np.tan(alpha))
yR = D / (t30 - np.tan(alpha))
zL = D + yL * np.tan(alpha)
zR = D + yR * np.tan(alpha)
ax.plot([yL, yR], [-zL, -zR], 'ro', ms=7, zorder=5)
ax.annotate('P_L', (yL, -zL), xytext=(yL - 45, -zL - 18),
            fontsize=11, color='red', fontweight='bold')
ax.annotate('P_R', (yR, -zR), xytext=(yR + 6, -zR - 18),
            fontsize=11, color='red', fontweight='bold')

# ===== 覆盖宽度 W (海平面基准, 贴在海平面上方) =====
ax.annotate('', xy=(yL, 12), xytext=(yR, 12),
            arrowprops=dict(arrowstyle='<->', color='red', lw=1.8))
ax.text((yL + yR) / 2, 24, f'W = {yR - yL:.1f} m',
        ha='center', color='red', fontsize=12, fontweight='bold')

# ===== 角度标注 (集中在船附近) =====
# 30°: 波束边缘与海平面的夹角 (数学角 -180°~-150°, 小弧)
ax.add_patch(Arc((0, 0), 30, 30, theta1=-180, theta2=-150, color='#2e75b6', lw=1.5))
ax.text(-30, -9, '30°', color='#2e75b6', fontsize=10,
        ha='center', va='center')
# 60°: 波束边缘与竖直向下方向的夹角 (数学角 -150°~-90°, 稍大弧)
ax.add_patch(Arc((0, 0), 46, 46, theta1=-150, theta2=-90, color='#2e75b6', lw=1.5))
ax.text(-23, -40, '60°', color='#2e75b6', fontsize=10,
        ha='center', va='center')
# 坡角 α (海底与水平线夹角, 紧贴坡面)
ax.add_patch(Arc((0, -D), 24, 24, theta1=-180, theta2=-180 + np.rad2deg(alpha),
                 color='#7030a0', lw=1.5))
ax.text(-14, -D + 5, f'α={np.rad2deg(alpha):.1f}°',
        color='#7030a0', fontsize=10, ha='center', va='center')

# ===== 水深标注 =====
ax.annotate('', xy=(12, 0), xytext=(12, -D),
            arrowprops=dict(arrowstyle='<->', color='k', lw=1.2, alpha=0.7))
ax.text(17, -D / 2, f'D={D}m', fontsize=10, va='center')

ax.set_xlim(-165, 165)
ax.set_ylim(-110, 60)
ax.set_xlabel('y (m)  向坡下为正', fontsize=11)
ax.set_ylabel('z (m)  深度(向下)', fontsize=11)
ax.legend(loc='lower right', fontsize=10)
ax.grid(True, alpha=0.2)
plt.tight_layout()
plt.savefig('D:/虚拟C盘/数学建模培训/单刷2023B/fig/fig_geometry.png',
            dpi=300, bbox_inches='tight')
print('[OK] output/fig_geometry.png')
